### ISSUE TITLE
Run tasks in parallel with `observe_on`

### ISSUE THREAD

**Issue by AntonOellerer** [rxRust/rxRust#211]
Hey,
While further developing with the library, I noticed behaviour which I did not expect from the documentation:
```Rust
  #[test]
  fn should_run_map_in_parallel() {
    let pool = ThreadPool::new().unwrap();
    observable::from_iter(1..10)
      .observe_on(pool.clone())
      .map(|v| {
        thread::sleep(Duration::from_secs(1));
        println!(
          "Mapping {} on {:?} at {:?}",
          v,
          thread::current().id(),
          SystemTime::now()
        );
        v
      })
      .into_shared()
      .subscribe(|v| {
        let handle = thread::current();
        println!(
          "Received {} on {:?} at {:?}",
          v,
          handle.id(),
          SystemTime::now()
        )
      });
    thread::sleep(Duration::from_secs(20));
  }
```

While one would expect that the tasks in the `map` function would run in parallel, I observe the following output:
```
Mapping 1 on ThreadId(3) at SystemTime { tv_sec: 1672906363, tv_nsec: 227367839 }
Received 1 on ThreadId(3) at SystemTime { tv_sec: 1672906363, tv_nsec: 227382646 }
Mapping 2 on ThreadId(4) at SystemTime { tv_sec: 1672906364, tv_nsec: 227563828 }
Received 2 on ThreadId(4) at SystemTime { tv_sec: 1672906364, tv_nsec: 227578774 }
Mapping 3 on ThreadId(5) at SystemTime { tv_sec: 1672906365, tv_nsec: 227742914 }
Received 3 on ThreadId(5) at SystemTime { tv_sec: 1672906365, tv_nsec: 227777766 }
Mapping 4 on ThreadId(7) at SystemTime { tv_sec: 1672906366, tv_nsec: 228213876 }
Received 4 on ThreadId(7) at SystemTime { tv_sec: 1672906366, tv_nsec: 228274220 }
Mapping 5 on ThreadId(6) at SystemTime { tv_sec: 1672906367, tv_nsec: 228547805 }
Received 5 on ThreadId(6) at SystemTime { tv_sec: 1672906367, tv_nsec: 228563799 }
Mapping 6 on ThreadId(8) at SystemTime { tv_sec: 1672906368, tv_nsec: 228751755 }
Received 6 on ThreadId(8) at SystemTime { tv_sec: 1672906368, tv_nsec: 228787096 }
Mapping 7 on ThreadId(9) at SystemTime { tv_sec: 1672906369, tv_nsec: 228991885 }
Received 7 on ThreadId(9) at SystemTime { tv_sec: 1672906369, tv_nsec: 229009486 }
Mapping 8 on ThreadId(10) at SystemTime { tv_sec: 1672906370, tv_nsec: 229211760 }
Received 8 on ThreadId(10) at SystemTime { tv_sec: 1672906370, tv_nsec: 229229430 }
Mapping 9 on ThreadId(11) at SystemTime { tv_sec: 1672906371, tv_nsec: 229670988 }
Received 9 on ThreadId(11) at SystemTime { tv_sec: 1672906371, tv_nsec: 229689636 }
```

Did I make a mistake here, or is there some issue in the `observe_on` implementation?

**Comment by AntonOellerer** [comment]
Examining it, thie reason seems to be in [observe_on.rs#L50](https://github.com/rxRust/rxRust/blob/0ddd72cb2ae902c26dee806770add0a9993a7097/src/ops/observe_on.rs#L50).
Even if the line is exectued on different threads per `Item`, each exeution waits on the completion of the previous `observer.next(v)`.
Sketched, the following modification:
```Rust
  fn next(&mut self, value: Self::Item) {
    self.observer_schedule(
      move |mut observer, v| {
        dbg!(thread::current().id());
        observer.next(v);
        dbg!(thread::current().id());
      },
      value,
    )
  }
```
produces the following output:
```
[src/ops/observe_on.rs:77] thread::current().id() = ThreadId(
    4,
)
[src/ops/observe_on.rs:77] thread::current().id() = ThreadId(
    5,
)
[src/ops/observe_on.rs:77] thread::current().id() = ThreadId(
    8,
)
[src/ops/observe_on.rs:77] thread::current().id() = ThreadId(
    3,
)
[src/ops/observe_on.rs:77] thread::current().id() = ThreadId(
    7,
)
Mapping #2 0 on ThreadId(4) at SystemTime { tv_sec: 1672752199, tv_nsec: 355831718 }
Received 0 on ThreadId(4) at SystemTime { tv_sec: 1672752200, tv_nsec: 355959854 }
[src/ops/observe_on.rs:79] thread::current().id() = ThreadId(
    4,
)
Mapping #2 2 on ThreadId(8) at SystemTime { tv_sec: 1672752201, tv_nsec: 356124517 }
Received 2 on ThreadId(8) at SystemTime { tv_sec: 1672752202, tv_nsec: 356266272 }
[src/ops/observe_on.rs:79] thread::current().id() = ThreadId(
    8,
)
Mapping #2 1 on ThreadId(5) at SystemTime { tv_sec: 1672752203, tv_nsec: 356488836 }
Received 1 on ThreadId(5) at SystemTime { tv_sec: 1672752204, tv_nsec: 356618717 }
[src/ops/observe_on.rs:79] thread::current().id() = ThreadId(
    5,
)
Mapping #2 3 on ThreadId(3) at SystemTime { tv_sec: 1672752205, tv_nsec: 356766618 }
Received 3 on ThreadId(3) at SystemTime { tv_sec: 1672752206, tv_nsec: 356897827 }
[src/ops/observe_on.rs:79] thread::current().id() = ThreadId(
    3,
)
Mapping #2 4 on ThreadId(7) at SystemTime { tv_sec: 1672752207, tv_nsec: 357089729 }
Received 4 on ThreadId(7) at SystemTime { tv_sec: 1672752208, tv_nsec: 357192093 }
[src/ops/observe_on.rs:79] thread::current().id() = ThreadId(
    7,
)
```

It seems if as some synchronization between each invocation takes place, causing the threads to synchronize

**Comment by AntonOellerer** [comment]
This is in turn caused by `self.observer` being wrapped in `MutArc`, which implements `rc_deref_mut` as `self.0.lock().unwrap()`, so blocking.

**Comment by M-Adoo** [comment]
I use `MutArc` wrapper to accept any downstream can be shared between threads. Sorry,  I have no idea to fix it in short time.  Because  we can't  require the `Observer` satisfy `Clone` in `LocalObservable`/`SharedObservable`.

But I have plan to refactor this library, the major work is merge the local & shared version traits. The problem will be resolved.

**Comment by AntonOellerer** [comment]
Ty!

Is there a way to parallelize long running tasks?
As in such that here:
```Rust
from_iter(0..10)
    .map(|v| {thread::sleep(Duration::from_secs(1); v}
    .subscribe(|v| println!("{v}");
```
each wait happens in another thread and thus a full run takes 1 second instead of 10?

**Comment by AntonOellerer** [comment]
What I think is happening:
1. `Thread A`:  `group_by` emits a `subject` to the `flat_map` ([group_by.rs#L63](https://github.com/rxRust/rxRust/blob/0ddd72cb2ae902c26dee806770add0a9993a7097/src/ops/group_by.rs#L63))
2. `Thread A`: `flat_map` requests to subscribe to `subject` ([flatten.rs#188](https://github.com/rxRust/rxRust/blob/0ddd72cb2ae902c26dee806770add0a9993a7097/src/ops/flatten.rs#L188))
2. `Thread A`: `group_by` emits value to `subject` ([group_by.rs#L66](https://github.com/rxRust/rxRust/blob/0ddd72cb2ae902c26dee806770add0a9993a7097/src/ops/group_by.rs#L66))
3. `Thread A`: `Subject` scans its list of `observers` which is still empty, discards value ([subject.rs#L234](https://github.com/rxRust/rxRust/blob/0ddd72cb2ae902c26dee806770add0a9993a7097/src/subject.rs#L234))
4. `Thread B`: `InnerSubject` adds the `flat_map` to the `observers` list ([subject.rs#L281](https://github.com/rxRust/rxRust/blob/0ddd72cb2ae902c26dee806770add0a9993a7097/src/subject.rs#L281))

The reason the test case is green sometimes is when a context switch from `Thread A` to `Thread B` happens, and by this moving `step 5` to after `step 2` but before `step 4`

**Comment by AntonOellerer** [comment]
One way to do it is like this:
```Rust
let pool_scheduler = ThreadPool::new().unwrap();
from_iter(0..5)
  .flat_map(move |v| {
    of(v)
      .tap(|v| thread::sleep(Duration::from_secs(1)))
      .subscribe_on(pool_scheduler.clone())
  })
  .into_shared()
  .subscribe(|v| println!("{v}: {:?}", SystemTime::now()));
```
but when building on this, issues such as the one mentioned in https://github.com/rxRust/rxRust/pull/210#issuecomment-1370633585 might arise

**Comment by AntonOellerer** [comment]
I will continue the conversation from https://github.com/rxRust/rxRust/pull/210 here, as it is better suited for this issue.

Upon further examination, I do not think that there really is a bug in `subscribe_on`, it's just a scheduling issue, which could maybe be solved by [yielding](https://doc.rust-lang.org/std/thread/fn.yield_now.html) the thread before each check of the subscribers queue ([subject.rs#L234](https://github.com/rxRust/rxRust/blob/0ddd72cb2ae902c26dee806770add0a9993a7097/src/subject.rs#L234))

Another point to consider when trying to parallelize long running tasks is that `subscribe_on` only invokes `actual_subscribe` on another thread, which requires a "cold" stream (emitting items upon subscription) instead of a "hot" one (this is also the reason why the combination does not interface properly with `group_by`/anything `Subject` based, as those do not perform the emission on `actual_subscribe`)

I think one way to solve this might be to rewrite `into_shared` such that it acts as an seperator between the asynchronous stream and the synchronous `subscribe` function, so that the [`.next()`](https://github.com/rxRust/rxRust/blob/0ddd72cb2ae902c26dee806770add0a9993a7097/src/ops/observe_on.rs#L50) calls do not block for each item.

I will have a look into that and report any findings here :)

**Comment by AntonOellerer** [comment]
Ok, from my further exploration it seems as if the solution to parallelize different runs of the same processing step transparently, one would have to remove the blocking <Arc> implementation with a solution instantiating `threads` and passing values via `channels` on each `next`, which would probably measure up to a rewrite of most of the sharing code.

But I don't know whether I understood everything correctly, more precisely how exactly `observe_on` works.
If it is not possible to parallelize computation of values, then what is the advantage of just calculating everything on one thread?

**Comment by M-Adoo** [comment]
The problem is  we use `MutArc`  wrapper downstream no matter if itself satisfy `Send`, `Sync`  or `Clone`, this force the emit value to downstream one by one.  

`observe_on`  should emit value by  scheduler,  if its scheduler support parallelize, it should be.  

So,  the most important things we need to refactor is if user want use a operator relate with "parallelize", its `subscribe` callback should satisfy the require(Send/Sync). Framework should not implicit use a wrapper to do it.  In this way,  the `shared` version traits should all be removed.

**Comment by M-Adoo** [comment]
@AntonOellerer   I fixed this issue in #68  and add a unit test. Is this what you want ?

https://github.com/rxRust/rxRust/blob/1395bc90249ced75429ed68c8a99df399003610d/src/ops/subscribe_on.rs#L89-L113

**Comment by AntonOellerer** [comment]
Hey,
Sorry for forgetting to answer!
I will take a look at it after my holiday, thank you!
In the meanwhile: I created a prototype of a message-passing implementation of the ReactiveX API, if you are curious you can take a look at it under https://github.com/AntonOellerer/rx_rust_mp :)
Best Regards,
Anton

**Comment by M-Adoo** [comment]
Seems you are not finished it.  I'm afraid I misled you to use channel to resolve the data race problem for `observe_on`. At that time, I didn't think deeply about this issue.  

- We should not avoid the data race of `observer_on`, because its semantic need pass the downstream observer to scheduler(maybe many threads) to emit value, so it must be thread safe and should forbid write it in same time. But that it what `next` need.  Although `observer_on` can't emit value in parallel, but it transform the work from current thread into scheduler, reduced the overhead of current thread.
- `subscribe_on` is what you want,  subscribe and process value in the scheduler.

**Comment by M-Adoo** [comment]
closed by #215
